lib/route/route_obj.c

```diff
@@ -1441,7 +1467,10 @@ int rtnl_route_build_msg(struct nl_msg *msg, struct rtnl_route *route)
 		nla_nest_end(msg, metrics);
 	}
 
-	if (rtnl_route_get_nnexthops(route) == 1) {
+	/* Nexthop specification and nexthop id are mutually exclusive */
+	if (route->rt_nhid > 0) {
```

Comment:
`if (mask & ROUTE_ATTR_NHID)`

Note that it is a questionable feature to let kernel choose the nexthop ID. However, if the user called `rtnl_route_set_nhid(route, 0)`, then that's what they should get.

Reply:
In the kernel RTA_NH_ID is stored in a (zeroed) struct, so 0 has here the meaning that there is no named nexthop, not that the kernel should choose/create one and assign an id.

All later checks are for `cfg->fc_nh_id > 0`. So for routes, RTA_NH_ID 0 has a slightly different meaning.

How about changing the setter to clear `ROUTE_ATTR_NHID` if being set to 0? Then it would mirror the meaning in the kernel, and we can use `ROUTE_ATTR_NHID` to mean it is set to a valid value. And you have a way of unsetting the value.

Reply:
> How about changing the setter to clear ROUTE_ATTR_NHID if being set to 0? Then it would mirror the meaning in the kernel, and we can use ROUTE_ATTR_NHID to mean it is set to a valid value. And you have a way of unsetting the value.

That sounds best!

Reply:
Done.